pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,248 @@
+package trace
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/tracer"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	defaultReqBufferSize      = 10000                  // tweaking this does not result in significant perf impact
+	DefaultBatchSize          = 5000                   // this is >= (we might produce bigger batches sometimes)
+	DefaultBatchTimeout       = 200 * time.Millisecond // arbitrary picked. We don't want client to wait too long
+	defaultMaxBufferedBatches = 200                    // arbitrary picked. We want to avoid wait on batches
+)
+
+var defaultBatchWorkers int
+
+func init() {
+	defaultBatchWorkers = runtime.NumCPU() / 2 // we only take half so other half can be used for writers
+	if defaultBatchWorkers == 0 {
+		// in case there <=1 CPU
+		defaultBatchWorkers = 1
+	}
+}
+
+// We batch individual InsertTracesReq
+type Batch struct {
+	traces    ptrace.Traces
+	spanCount int
+	reqStatus []chan error
+	ctx       context.Context
+	maxSize   int
+}
+
+func NewBatch(maxBatchSize int) *Batch {
+	return &Batch{
+		traces:    ptrace.NewTraces(),
+		spanCount: 0,
+		reqStatus: make([]chan error, 0, maxBatchSize),
+		ctx:       context.Background(),
+		maxSize:   maxBatchSize, // this is not hard limit
+	}
+}
+
+func (tb *Batch) add(in InsertTracesReq) {
+	_, addReqSpan := tracer.Default().Start(tb.ctx, "add-trace-req-to-batch",
+		trace.WithLinks(trace.Link{SpanContext: in.spanCtx}))
+	defer addReqSpan.End()
+	inSpans := in.payload.SpanCount()
+	if inSpans == 0 {
+		return
+	}
+	in.payload.ResourceSpans().MoveAndAppendTo(tb.traces.ResourceSpans())
+	tb.spanCount += inSpans
+	tb.reqStatus = append(tb.reqStatus, in.err)
+}
+
+func (tb *Batch) isFull() bool {
+	return tb.spanCount >= tb.maxSize
+}
+
+func (tb *Batch) isEmpty() bool {
+	return tb.spanCount == 0
+}
+
+// Batcher batches trace requests and sends batches to batch writer
+// This is done to achieve better ingest performance especially b/c
+// Jaeger collector sends traces one by one
+type Batcher struct {
+	in              chan InsertTracesReq
+	stop            chan struct{}
+	batchWriter     *batchWriter
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+type WriteTraces func(context.Context, ptrace.Traces) error
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	if config.Batchers == 0 {
```

Comment:
It might be better to move all these checks/validity to `verifyConfig(&config)`. Keeps the constructor clean.